{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e27593e1-56d7-4051-8180-8c00d6cb400f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tss\n",
    "import gzip\n",
    "\n",
    "gtf_file = \"gencode.v48.basic.annotation.gtf.gz\"\n",
    "output_bed = \"gencode_v48_TSSs.bed\"\n",
    "\n",
    "def parse_gtf_attributes(attr_str):\n",
    "    \"\"\"Extract gene_id and gene_name from GTF attributes column.\"\"\"\n",
    "    attrs = dict(item.strip().replace('\"','').split(' ', 1)\n",
    "                 for item in attr_str.strip().split(';') if item)\n",
    "    return attrs.get(\"gene_id\", \".\"), attrs.get(\"gene_name\", \".\")\n",
    "\n",
    "with gzip.open(gtf_file, 'rt') as infile, open(output_bed, 'w') as outfile:\n",
    "    for line in infile:\n",
    "        if line.startswith(\"#\"):\n",
    "            continue\n",
    "        cols = line.strip().split('\\t')\n",
    "        if cols[2] != \"transcript\":\n",
    "            continue\n",
    "\n",
    "        chrom = cols[0]\n",
    "        start = int(cols[3])\n",
    "        end = int(cols[4])\n",
    "        strand = cols[6]\n",
    "        attr = cols[8]\n",
    "\n",
    "        gene_id, gene_name = parse_gtf_attributes(attr)\n",
    "\n",
    "        # Determine TSS (0-based for BED)\n",
    "        tss = start - 1 if strand == '+' else end - 1\n",
    "\n",
    "        # Write to BED file (1 bp window at TSS)\n",
    "        bed_line = f\"{chrom}\\t{tss}\\t{tss+1}\\t{gene_name}\\t.\\t{strand}\\n\"\n",
    "        outfile.write(bed_line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "656c29d3-61d8-49e5-8e21-42609ccd9202",
   "metadata": {},
   "outputs": [],
   "source": [
    "#promoter annotations\n",
    "import gzip\n",
    "\n",
    "gtf_file = \"gencode.v48.basic.annotation.gtf.gz\"\n",
    "output_bed = \"gencode_v48_promoters.bed\"\n",
    "promoter_window = 1000  # Â±1 kb\n",
    "\n",
    "def parse_gtf_attributes(attr_str):\n",
    "    attrs = dict(item.strip().replace('\"','').split(' ', 1)\n",
    "                 for item in attr_str.strip().split(';') if item)\n",
    "    return attrs.get(\"gene_id\", \".\"), attrs.get(\"gene_name\", \".\")\n",
    "\n",
    "with gzip.open(gtf_file, 'rt') as infile, open(output_bed, 'w') as outfile:\n",
    "    for line in infile:\n",
    "        if line.startswith(\"#\"):\n",
    "            continue\n",
    "        cols = line.strip().split('\\t')\n",
    "        if cols[2] != \"transcript\":\n",
    "            continue\n",
    "\n",
    "        chrom = cols[0]\n",
    "        start = int(cols[3])\n",
    "        end = int(cols[4])\n",
    "        strand = cols[6]\n",
    "        attr = cols[8]\n",
    "\n",
    "        gene_id, gene_name = parse_gtf_attributes(attr)\n",
    "\n",
    "        tss = start if strand == '+' else end\n",
    "        promoter_start = max(0, tss - promoter_window)\n",
    "        promoter_end = tss + promoter_window\n",
    "\n",
    "        bed_line = f\"{chrom}\\t{promoter_start}\\t{promoter_end}\\t{gene_name}\\t.\\t{strand}\\n\"\n",
    "        outfile.write(bed_line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b6a1d57c-d0e1-434d-9b0e-fece361a56f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene bodies\n",
    "import gzip\n",
    "\n",
    "gtf_file = \"gencode.v48.basic.annotation.gtf.gz\"\n",
    "output_bed = \"gencode_v48_gene_bodies.bed\"\n",
    "\n",
    "def parse_gtf_attributes(attr_str):\n",
    "    attrs = dict(item.strip().replace('\"','').split(' ', 1)\n",
    "                 for item in attr_str.strip().split(';') if item)\n",
    "    return attrs.get(\"gene_id\", \".\"), attrs.get(\"gene_name\", \".\")\n",
    "\n",
    "with gzip.open(gtf_file, 'rt') as infile, open(output_bed, 'w') as outfile:\n",
    "    for line in infile:\n",
    "        if line.startswith(\"#\"):\n",
    "            continue\n",
    "        cols = line.strip().split('\\t')\n",
    "        if cols[2] != \"gene\":\n",
    "            continue\n",
    "\n",
    "        chrom = cols[0]\n",
    "        start = int(cols[3]) - 1  # BED is 0-based\n",
    "        end = int(cols[4])\n",
    "        strand = cols[6]\n",
    "        attr = cols[8]\n",
    "\n",
    "        gene_id, gene_name = parse_gtf_attributes(attr)\n",
    "\n",
    "        bed_line = f\"{chrom}\\t{start}\\t{end}\\t{gene_name}\\t.\\t{strand}\\n\"\n",
    "        outfile.write(bed_line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e212be5a-f6d6-47fd-b74d-de850002a8c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#exons\n",
    "import gzip\n",
    "\n",
    "gtf_file = \"gencode.v48.basic.annotation.gtf.gz\"\n",
    "output_bed = \"gencode_v48_exons.bed\"\n",
    "\n",
    "def parse_gtf_attributes(attr_str):\n",
    "    attrs = dict(item.strip().replace('\"','').split(' ', 1)\n",
    "                 for item in attr_str.strip().split(';') if item)\n",
    "    return attrs.get(\"gene_id\", \".\"), attrs.get(\"gene_name\", \".\")\n",
    "\n",
    "with gzip.open(gtf_file, 'rt') as infile, open(output_bed, 'w') as outfile:\n",
    "    for line in infile:\n",
    "        if line.startswith(\"#\"):\n",
    "            continue\n",
    "        cols = line.strip().split('\\t')\n",
    "        if cols[2] != \"exon\":\n",
    "            continue\n",
    "\n",
    "        chrom = cols[0]\n",
    "        start = int(cols[3]) - 1\n",
    "        end = int(cols[4])\n",
    "        strand = cols[6]\n",
    "        attr = cols[8]\n",
    "\n",
    "        gene_id, gene_name = parse_gtf_attributes(attr)\n",
    "        bed_line = f\"{chrom}\\t{start}\\t{end}\\t{gene_name}\\t.\\t{strand}\\n\"\n",
    "        outfile.write(bed_line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "81ec50e8-49a1-4187-8f3f-93591bae691d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#introns\n",
    "from collections import defaultdict\n",
    "\n",
    "gtf_file = \"gencode.v48.basic.annotation.gtf.gz\"\n",
    "output_bed = \"gencode_v48_introns.bed\"\n",
    "\n",
    "def parse_gtf_attributes(attr_str):\n",
    "    attrs = dict(item.strip().replace('\"','').split(' ', 1)\n",
    "                 for item in attr_str.strip().split(';') if item)\n",
    "    return attrs.get(\"transcript_id\", \".\"), attrs.get(\"gene_name\", \".\")\n",
    "\n",
    "# Collect exon coordinates per transcript\n",
    "transcripts = defaultdict(lambda: {'chrom': '', 'strand': '', 'exons': [], 'gene_name': ''})\n",
    "\n",
    "with gzip.open(gtf_file, 'rt') as infile:\n",
    "    for line in infile:\n",
    "        if line.startswith(\"#\"):\n",
    "            continue\n",
    "        cols = line.strip().split('\\t')\n",
    "        if cols[2] != \"exon\":\n",
    "            continue\n",
    "\n",
    "        chrom = cols[0]\n",
    "        start = int(cols[3]) - 1\n",
    "        end = int(cols[4])\n",
    "        strand = cols[6]\n",
    "        attr = cols[8]\n",
    "\n",
    "        tx_id, gene_name = parse_gtf_attributes(attr)\n",
    "\n",
    "        transcripts[tx_id]['chrom'] = chrom\n",
    "        transcripts[tx_id]['strand'] = strand\n",
    "        transcripts[tx_id]['gene_name'] = gene_name\n",
    "        transcripts[tx_id]['exons'].append((start, end))\n",
    "\n",
    "# Infer introns\n",
    "with open(output_bed, 'w') as out:\n",
    "    for tx_id, info in transcripts.items():\n",
    "        chrom = info['chrom']\n",
    "        strand = info['strand']\n",
    "        gene_name = info['gene_name']\n",
    "        exons = sorted(info['exons'])\n",
    "\n",
    "        for i in range(len(exons) - 1):\n",
    "            intron_start = exons[i][1]\n",
    "            intron_end = exons[i + 1][0]\n",
    "            if intron_start < intron_end:\n",
    "                out.write(f\"{chrom}\\t{intron_start}\\t{intron_end}\\t{gene_name}\\t.\\t{strand}\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d1645f37-8112-405d-b9f8-0fed011e9fae",
   "metadata": {},
   "outputs": [],
   "source": [
    "#biotype protein_coding, lncrna\n",
    "\n",
    "gtf_file = \"gencode.v48.basic.annotation.gtf.gz\"\n",
    "output_bed = \"gencode_v48_gene_biotypes.bed\"\n",
    "\n",
    "def parse_gtf_attributes(attr_str):\n",
    "    attrs = {}\n",
    "    for item in attr_str.strip().split(';'):\n",
    "        if item:\n",
    "            key, val = item.strip().split(' ', 1)\n",
    "            attrs[key] = val.strip('\"')\n",
    "    return attrs\n",
    "\n",
    "with gzip.open(gtf_file, 'rt') as infile, open(output_bed, 'w') as outfile:\n",
    "    for line in infile:\n",
    "        if line.startswith(\"#\"):\n",
    "            continue\n",
    "        cols = line.strip().split('\\t')\n",
    "        if cols[2] != \"gene\":\n",
    "            continue\n",
    "\n",
    "        chrom = cols[0]\n",
    "        start = int(cols[3]) - 1  # BED is 0-based\n",
    "        end = int(cols[4])\n",
    "        strand = cols[6]\n",
    "        attr = parse_gtf_attributes(cols[8])\n",
    "\n",
    "        gene_name = attr.get(\"gene_name\", \".\")\n",
    "        gene_type = attr.get(\"gene_type\", attr.get(\"gene_biotype\", \".\"))\n",
    "\n",
    "        bed_line = f\"{chrom}\\t{start}\\t{end}\\t{gene_name}|{gene_type}\\t.\\t{strand}\\n\"\n",
    "        outfile.write(bed_line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "347b0be5-92b0-48a6-a76d-5d3cd4c9ce56",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene annotations\n",
    "import gzip\n",
    "\n",
    "gtf_file = \"gencode.v48.basic.annotation.gtf.gz\"\n",
    "output_bed = \"gencode_v48_gene_symbols_ids.bed\"\n",
    "\n",
    "def parse_gtf_attributes(attr_str):\n",
    "    attrs = {}\n",
    "    for item in attr_str.strip().split(';'):\n",
    "        if item:\n",
    "            key, val = item.strip().split(' ', 1)\n",
    "            attrs[key] = val.strip('\"')\n",
    "    return attrs\n",
    "\n",
    "with gzip.open(gtf_file, 'rt') as infile, open(output_bed, 'w') as outfile:\n",
    "    for line in infile:\n",
    "        if line.startswith(\"#\"):\n",
    "            continue\n",
    "        cols = line.strip().split('\\t')\n",
    "        if cols[2] != \"gene\":\n",
    "            continue\n",
    "\n",
    "        chrom = cols[0]\n",
    "        start = int(cols[3]) - 1  # BED format: 0-based\n",
    "        end = int(cols[4])\n",
    "        strand = cols[6]\n",
    "        attr = parse_gtf_attributes(cols[8])\n",
    "\n",
    "        gene_id = attr.get(\"gene_id\", \".\")\n",
    "        gene_name = attr.get(\"gene_name\", \".\")\n",
    "\n",
    "        bed_line = f\"{chrom}\\t{start}\\t{end}\\t{gene_name}|{gene_id}\\t.\\t{strand}\\n\"\n",
    "        outfile.write(bed_line)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e839116-5a40-48b6-9b22-1ebdd3cc5e60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "78686\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "with open(\"gencode_v48_gene_symbols_ids.bed\") as f:\n",
    "    for line in f:\n",
    "        if line.strip():  # skip empty lines\n",
    "            count += 1\n",
    "print(count)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
